lib/AST/UnqualifiedLookup.cpp

```diff
@@ -343,8 +344,60 @@ void UnqualifiedLookupFactory::ResultFinderForTypeContext::findResults(
   SmallVector<ValueDecl *, 4> Lookup;
   contextForLookup->lookupQualified(selfBounds, Name, baseNLOptions, Lookup);
   for (auto Result : Lookup) {
-    results.emplace_back(const_cast<DeclContext *>(whereValueIsMember(Result)),
-                         Result);
+    auto baseDC = const_cast<DeclContext *>(whereValueIsMember(Result));
+    
+    // Retrieve the base decl for this lookup
+    ValueDecl* baseDecl;
+    
+    // Perform an unqualified lookup for the base decl of this result. This handles cases
+    // where self was rebound (e.g. `guard let self = self`) earlier in the scope.
+    //  - Only do this in closures that capture self weakly, since implicit self isn't
+    //    allowed to be rebound in other contexts. In other contexts, implicit self
+    //    _always_ refers to the context's self `ParamDecl`, even if there is another
+    //    local decl with the name `self` that would be found by `lookupSingleLocalDecl`.
+    ValueDecl* localBaseDecl = nullptr;
+    bool isInWeakSelfClosure = false;
+    if (auto closureExpr = dyn_cast<ClosureExpr>(factory->DC))
+      if (auto selfDecl = closureExpr->getCapturedSelfDecl())
+        if (auto *attr = selfDecl->getAttrs().getAttribute<ReferenceOwnershipAttr>())
+          if (attr->get() == ReferenceOwnership::Weak)
+            isInWeakSelfClosure = true;
+            
+    if (isInWeakSelfClosure) {
+      localBaseDecl = ASTScope::lookupSingleLocalDecl(factory->DC->getParentSourceFile(),
+                                                      DeclName(factory->Ctx.Id_self),
+                                                      factory->Loc);
+    }
+    
+    if (baseDC == nullptr)
```

Comment:
Let's package this into a closure just like in other place, this would reduce nesting and remove need for `else if`s here. Also please note that we use `clang-format` to format changes  before commit - https://github.com/apple/swift/blob/main/docs/HowToGuides/FAQ.md#how-do-i-format-my-changes